### DISCUSSION TITLE
Lesson 4, video 8, DP example of ages. The calculation might be wrong.

### DISCUSSION THREAD

**Discussion by gonzalo-munillag** [OpenMined/courses#144] [Course: Our Privacy Opportunity]
Andrew explains a re-identification attack based on the mean ages of poll participants, Andrew says that the victim's age is 20, but given the data he provides:

x/1000 = 27

y/999 = 26.98

x = y + z

Solving for z (age of the victim) gives 46.98.

Perhaps I am wrong in my logic, it would be great if you could talk about it :)
However counterintuitive, it makes sense, if the average is smaller than before, that means that the missing age is larger. 
Or perhaps he just wanted to make a point, which was very well made.

Thank you in advance!
Cheers

**Comment by ganevgv** [comment]
Good catch, Gonzalo! I'm getting the same result as you :)

**Comment by yemikifouly** [comment]
@mcleonard Could you please help with this question?

**Comment by ZSoumia** [comment]
@gonzalo-munillag When we re-calculate the average of ages without the one being taken out we will be calculating the same average taking into consideration that the number of people in the group are still 1000 and not 999 ( you are taking off only the paticipation of that individual and not entirely changing the group and considering him/her out ). So: 
avg1 = ( x<sub>1</sub>+x<sub>2</sub>+.........+x<sub>1000</sub>) / 1000 = 27 
So (x<sub>1</sub> + x<sub>2</sub> +.........+ x<sub>1000</sub> ) = 27 * 1000 = 27000
Let's assume the individual you didn't send him the request to re-run the survey is x<sub>1</sub> ( remember you didn't ask for his/her input but that individual is still a part of the group) 
avg2 = (x<sub>2</sub> +.........+ x<sub>1000</sub>) / 1000 = 26.98 
(x<sub>2</sub>+.........+x<sub>1000</sub>) = 26.98 * 1000 = 26980
and hence: 
x<sub>1</sub> =  27000 - 26980 = 20

**Reply by gonzalo-munillag** [reply]
Great, thank you @ZSoumia! It is awesome that something apparently simple has its intricacies :)

While it makes sense, I still do not see clearly using 1000 again. "you are taking off only the participation of that individual and not changing the group", but why "taking off the participation" only means the number in the numerator and not also a unit in the denominator? Your argument, at least for me, could apply to many interpretations of what you can take off or not from the group. "He is still part of the group" - Why does that only apply to the numerator?
Moreover, avg2 is inherently not correct, you only ask 999 individuals, therefore, avg2 is not representing the 999 people.
Additionally, could you explain why the approach I presented is conceptually wrong?

Is there another way to explain this conundrum or a source where it is proven that using only 1000 is the right way to go?

Thank you so much!! This discussion is awesome, loved the course.
Cheers!

**Reply by mmalekzadeh** [reply]
@gonzalo-munillag:  I think you are right.

This in general is called a "differencing attack" and in this attack, we sometimes even don't need to know the number of people, but only to know that the difference between the number of people in the two queries is exactly 1. 

So, we receive 27 for query 1 and 26.98 for query 2, and we know that query 2 does not include the victim's data.
So, if the absence of the victim makes the average smaller, then the victim must be older than the average.

@ZSoumia: Thanks for your explanation. Could you please explain what exactly do you mean by this: "remember you didn't ask for his/her input but that individual is still a part of the group"?. If you query the system "what is the average of these 999 people", why does the system should divide the sum by 1000? Consider that in the example explained in the video, the attacker knows that 999 people are asked.

**Reply by ganevgv** [reply]
@mmalekzadeh good observation. If we are indeed talking about a differencing attack in this example, Andrew explains it in more detail in his [Secure and Private AI](https://www.udacity.com/course/secure-and-private-ai--ud185) course, [Lesson 3: Evaluating the Privacy of a Function, Concept 7. Project Demo Perform A Differencing Attack](https://classroom.udacity.com/courses/ud185/lessons/3879a8dd-9393-4c7f-a615-5ac1bdb2a629/concepts/5bacf26d-0ae7-4218-8ffe-55a9fca8c0e0). There, the function (in our case mean) is applied to two neighbouring (defined also as parallel) databases - one with you and one without you. In both cases, the actual number or participants in the query is used:
<img width="596" alt="Screenshot 2021-01-08 at 00 57 42" src="https://user-images.githubusercontent.com/17522184/103954332-a6a30d00-514c-11eb-96aa-ba448f4162c3.png">
In our example, that would be 1000 in the first query and 999 in the second.

**Reply by gonzalo-munillag** [reply]
@ganevgv Thanks a lot for looking that up. I think the code snippet clears our questions to some extent; Andrew was probably improvising and used his intuition, but the attack result is counter-intuitive. The important thing is that his point did come across correctly. And thanks @mmalekzadeh for participating in the discussion as well!

**Comment by manik-500** [comment]
Hi @gonzalo-munillag @ganevgv 
Let's make the calculation simple .
Suppose we want  average age of 4 people rather than 1000

If the 4 people age are 
p1=20,
p2=25
p3=30
p4=35
Then their average age will be (20+25+30+40)/4=27.5

For the second case Andrew rerun the survey except one
Suppose p4 didn’t participate in the second survey .
So their average age will be (20+25+30+0)/4=18.75 

Note: I think the survey app put default value 0 (zero ) when someone didn’t participate in the survey and counted total people who were included in the survey rather than who participated.

So as 4 people participate in the first survey then total age of 4 people will be (27.5 * 4)=110
So as 3 people participate in the second survey then total age of 3 people will be (18.75 * 4)=75

Note : we multiply 18.75 by 4 because when averaging for the second case we divide the total by 4

So 4th people age will be 110-75=35

If you apply this above procedure for 1000 people then the calculation will be 

Total age of 1000 people will be (27*1000) =27000
Total age of 999 people will be (26.98 *1000) =26980
So the remaining person age is (27000-26980)=20

**Reply by leriomaggio** [reply]
To summarise, IMHO both @gonzalo-munillag and @manik-500 are right and the very difference is whether the total number of items in the average calculation changes or not in the second case (as in `1,000` vs `999`).

Either way, the conclusion is the same: **it is possible to disclose the missing piece of information** (only the method to calculate it varies, but the principle stands :)

As for the error in the video: I personally had the same concern, but I do agree with @manik-500's interpretation so the video does not actually contain a mistake :)

my2¢

**Reply by manik-500** [reply]
I agree with @leriomaggio that by re-identification attack we can disclose missing information .
And that is the main point Andrew tried to explain in the video .
But The outcome varies here because I think for the second survey , counted total people as who were included in the survey rather than who participated .
On the other hand @gonzalo-munillag and @ganevgv counted total people as who participated rather than who were included in the survey .
So I think both calculation is right based on their assumption .
